_posts/2023-04-14-kafka-transactions.md

```diff
@@ -0,0 +1,340 @@
+
+If you've not collected the metrics then you have no information to go on when something goes wrong. 
+You're often left with no choice but to reconfigure with the right metrics and wait for the problem to happen again in order to understand it. 
+This could double its impact, or worse.
+
+On the broker side, the following JMX metrics are particularly helpful when diagnosing issue impacting transactions.
+
+```sh
+# requests: FindCoordinator, InitProducerId, AddPartitionsToTxn, Produce, EndTxn
+kafka.network:type=RequestMetrics,name=RequestsPerSec,request=([\w]+)
+kafka.network:type=RequestMetrics,name=RequestQueueTimeMs,request=([\w]+)
+kafka.network:type=RequestMetrics,name=TotalTimeMs,request=([\w]+)
+kafka.network:type=RequestMetrics,name=LocalTimeMs,request=([\w]+)
+kafka.network:type=RequestMetrics,name=RemoteTimeMs,request=([\w]+)
+kafka.network:type=RequestMetrics,name=ThrottleTimeMs,request=([\w]+)
+kafka.network:type=RequestMetrics,name=ResponseQueueTimeMs,request=([\w]+)
+kafka.network:type=RequestMetrics,name=ResponseSendTimeMs,request=([\w]+)
+# attributes: Count, 99thPercentile
+```
+
+Likewise the following producer metrics can be helpful, perhaps after broker logs reveal some misbehavior.
+
+```sh
+kafka.producer:type=producer-metrics,client-id=([-.\w]+)
+# attributes: txn-init-time-ns-total, txn-begin-time-ns-total, txn-send-offsets-time-ns-total, txn-commit-time-ns-total, txn-abort-time-ns-total
+```
+
+### Drawbacks
+
+There are a few drawbacks when using Kafka transactions that developers should be aware of.
+
+Firstly, enabling EOS has an effect on throughput, which can be significant.
+Each producer can only process transactions sequentially, so you might need multiple transactional producers, which can have a knock-on effect on the rest of the application.
+This is where frameworks can provide useful abstractions over multiple producer instances.
+
+Then, transactions can add some latency to an application due to the write amplification required to ensure EOS.
+While this may not be a significant issue for some applications, those that require real-time processing may be affected.
+
+Each transaction in Kafka incurs some overhead that is independent of the number of messages involved. 
+This overhead includes the following:
+
+- A small number of Remote Procedure Calls (RPCs) to register the partitions with the coordinator. 
+  These RPCs can be batched together to reduce overhead.
+- A marker that must be written to each partition when the transaction is completed.
+- A few state change records that are appended to the internal transaction state log.
+
+Most of this latency is on the producer side, where the transactions API is implemented.
+The performance degradation can be significant when having short transaction commit intervals.
+Increasing the transaction duration also increases the end-to-end latency and may require some additional tuning, so it's a matter of tradeoff.
+
```

Comment:
It's worth mentioning the impacts on the `isolation.level=read_committed` consumer too. Although it will only be able to fetch upto the LSO it still has to buffer records in memory until is observes the commit marker, so you expect to see increased memory usage.

Reply:
That's true. Thanks for pointing this out.